accepted/install-locations.md

```diff
@@ -0,0 +1,239 @@
+* `./host/fxr/<version>/hostfxr.dll` - this library contains all the logic to resolve framework references and select SDK. Regardless of application, the latest available version is used.
+* `./shared` - contains shared frameworks.
+  * `./Microsoft.NETCore.App/<version>` - the root framework which contains the .NET Core runtime. This framework contains `hostpolicy.dll` which is the entry point called by the `hostfxr` to run the app. Applications can choose which version to use.
+  * `./<FrameworkName>/<version>` - any other shared framework. Applications can choose which version to use.
+* `./sdk/<version>` - only installed when .NET Core SDK is installed - contains the SDK tools. Users can choose which version to use.
+
+
+## Installation types
+.NET Core can be installed in several ways, each provides a different location and expected behavior. Each such location will contain the .NET Core install layout described above.
+
+### Global install to default location
+This is the most common case where an installer installs the runtime (and potentially SDK) to a global default location accessible to everything on the system (and by all users).
+* Windows
+  * 32-bit on 64-bit OS - `%ProgramFiles(x86)%\dotnet`
+  * Otherwise - `%ProgramFiles%\dotnet`
+* Linux - the host searches in `/usr/share/dotnet` as a fallback, but the installer may pick a different default based on the requirements of the specific distro/configuration. In such case the install is effectively just like the global install to custom location described below. If correctly registered the host will search the custom location first.
+* macOS - `/usr/local/share/dotnet`
+
+Typically the muxer from this install is registered to be readily accessible without specifying full path. The exact mechanism is platform specific, for example on Windows this means that the muxer is added to the `PATH`.
+
+*This type of install is supported on all versions of .NET Core.*
+
+### Global install to custom location
+.NET Core is installed globally for the machine but into a non-default (custom) location. The location of the install is stored in platform specific configuration so that it can be found by everything on the system (and by all users).
+
+This type of install should effectively behave the same as install to the default location. Can be used to put .NET Core on a different drive to save space on the system drive, or to serve other purposes.
+Typically the muxer from this install is registered to be readily accessible without specifying full path. The exact mechanism is platform specific, for example on Windows this means that the muxer is added to the `PATH`.
+
+Installer behavior:
+* There can only be one globally registered location.
+* The installer should query the system for existing .NET Core global installs. If it finds any, it should only install to the already registered location, it should not install to a new location.
```

Comment:
When you say 'installer' here, does that include Linux packages? Do package managers allow this in general? I ask because RPM, to use an example, decides the list of installation paths at RPM-build time. At install time, the files simply get installed to that path selected at RPM-build time. Detecting and using different paths is not the common workflow and will result in inconsistencies.

Reply:
I'm not a Linux package manager expert by any means, so if you have a better idea, I would be very interested. 

In general the scenario is like:
* Application is executed via `apphost` with no context (nothing interesting in environment variables and so on)
* It needs to be able to find the "glocal install location"
* It goes to the config file and reads it from there.

The config file is not versioned in any way (.NET Core 3 apps will not have a different config file from .NET Core 3.1 apps). So the "installer" (whatever it is) must somehow figure out a single location where to put files for all versions.

It would be possible to modify the host to somehow support multiple locations based on some kind of version, but this is really difficult, especially in the world of multiple shared frameworks. It is very unlikely we will be able to build something like this for .NET Core 3 (even if we wanted to).

Potentially the config file could support a list of install locations (and the host would search them all for the requested frameworks). This would obviously be prone to ordering issues if the same version is installed multiple times. It also has a potential to degrade performance if there are too many such locations. The other problem is that the host is split into two parts `hostfxr` and `hostpolicy`. `hostfxr` is the part which makes the versioning decisions, as such it in itself is not really versioned (one is used for all apps) - if there are multiple global install locations, how would we find the "best one" for a given app. Some of this is basically discussed in the document [here](https://github.com/dotnet/designs/blob/be73a80791477ce6f1bbd7b23a0df18c60fe4380/accepted/install-locations.md#hostfxr-search-1)

If anybody knows what is the Linux way of solving a scenario like this, I would be very interested in learning more.

Reply:
I think you should just say that this bullet point about installer behavior applies to Windows only.

On Linux, the distro builders can pick the one location for given distro and this location will be written to `/etc/dotnet`.